### PR TITLE
feat: Allow setting context to null from beforeSend

### PR DESCRIPTION
**Pull request by kattrali**
## Changeset

* Moved automatic context detection from during serialization to capture-time, which is in `Client`
* Deleted a test for the context logic in Error

## Tests

* [x] Add new test scenario for overriding context
* (Other scenarios cover automatic context assignment and setting context to a new value)

## Review

- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## TBD
+
+### Bug fixes
+
+* Allow setting context to null from callbacks
+  [#380](https://github.com/bugsnag/bugsnag-android/pull/380)
```
Comment (fractalwrench) [inline CHANGELOG.md]: The PR should be 381

File: sdk/src/main/java/com/bugsnag/android/Client.java [hunk 2]
```diff
@@ -838,6 +838,10 @@ void notify(@NonNull Error error,
         // Attach user info to the error
         error.setUser(user);
 
+        // Attach default context from active activity
+        String context = config.getContext();
+        error.setContext(context != null ? context : appData.getActiveScreenClass());
```
Comment (fractalwrench) [inline sdk/src/main/java/com/bugsnag/android/Client.java]: This needs to check whether the context has already been set on the error object, and avoid setting the field if this is the case. For example, the context will be wrong for the following `notify` method on `Client`:

```
Bugsnag.getClient().notify("Test", "exc_message", "CustomContext", new StackTraceElement[]{}, Severity.ERROR, new MetaData());
```

File: sdk/src/androidTest/java/com/bugsnag/android/ErrorTest.java [hunk 3]
```diff
@@ -314,28 +314,13 @@ public void testBugsnagExceptionName() throws Exception {
         assertEquals("Busgang", err.getExceptionName());
     }
 
-    @Test
-    public void testConfigContext() throws Exception {
-        String expected = "Junit test suite";
-        error.setContext(null);
-        config.setContext(expected);
```
Comment (fractalwrench) [inline sdk/src/androidTest/java/com/bugsnag/android/ErrorTest.java]: This test could be reworked so that `config.setContext` is called before the error is built.
